staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package responsewriter
+
+import (
+	"bufio"
+	"net"
+	"net/http"
+)
+
+// UserProvidedDecorator represensts a user (client that uses this package)
+// provided decorator that wraps an inner http.ResponseWriter object.
+// The user provided decorator object must return the inner (decorated)
+// http.ResponseWriter object via the Unwrap function.
+type UserProvidedDecorator interface {
+	http.ResponseWriter
+
+	// Unwrap returns the inner http.ResponseWriter object associated
+	// with the user provided decorator.
+	Unwrap() http.ResponseWriter
+}
+
+// WrapForHTTP1Or2 accepts a user provided decorator of an "inner" http.responseWriter
+// object and potentially wraps it with a new http.ResponseWriter object that
+// implements http.CloseNotifier, http.Flusher, and/or http.Hijacker by
+// delegating to the user provided decorator or the inner
+// http.ResponseWriter, so that the returned http.ResponseWriter object
+// implements the same subset of those interfaces as the inner http.ResponseWriter.
+//
+// This function handles the following three casses.
+// - The inner ResponseWriter implements `http.CloseNotifier`, `http.Flusher`,
+//   and `http.Hijacker` (an HTTP/1.1 sever provides such a ResponseWriter).
+// - The inner ResponseWriter implements `http.CloseNotifier` and `http.Flusher`
+//   but not `http.Hijacker` (an HTTP/2 server provides such a ResponseWriter).
+// - All the other cases collapse to this one, in which the given ResponseWriter is returned.
+//
+// There are three applicable terms:
+//  - "outer": this is the ResponseWriter object returned by the WrapForHTTP1Or2 function.
+//  - "user provided decorator" or "middle": this is the user provided decorator
+//     that decorates an inner ResponseWriter object. A user provided decorator
+//     implements the UserProvidedDecorator interface. A user provided decorator
+//     may or may not implement http.CloseNotifier, http.Flusher or http.Hijacker.
+//  - "inner": the inner object may be the original ResposneWriter object passed
+//     to the first request handler in the filter chain or could be an "outer"
+//     ResponseWriter object if WrapForHTTP1Or2 has been invoked multiple times.
```

Comment:
I think this is too specific.  This function works with any inner ResponseWriter.  What is important here is what _this_ function does, which is documented earlier in the comment.

Reply:
Perhaps simply write something like the following here.

> // - "inner": the ResponseWriter that the user-provided decorator extends.